### PR TITLE
[Halmos] Setup for verification and first invariants

### PR DESCRIPTION
**Pull request by QGarchery**
Todo:
- [x] use test profile for Halmos
- [x] remove installation of setuptools when no longer necessary
- [x] ~~look at certora timeout invariants~~ (all of them are timing out because of non linear math, which Halmos can't help with)
- [x] look at certora that could be quick win invariants

### CODE REVIEW

File: test/forge/HalmosTest.sol
Comment (MathisGD) [inline test/forge/HalmosTest.sol:1]: also, when running locally i get this

$${\color{Skipped ConstantsLib.json due to parsing failure: KeyError: 'metadata'}Red}$$

File: test/forge/HalmosTest.sol [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        vm.stopPrank();
+
+        vm.roll(block.number + 1);
+        vm.warp(block.timestamp + 1 * BLOCK_TIME);
+    }
+
+    // Call Morpho, assuming interacting with only the defined market for performance reasons.
+    function _callMorpho(bytes4 selector, address caller) internal {
+        vm.assume(selector != morpho.extSloads.selector);
+        vm.assume(selector != morpho.createMarket.selector);
```
Comment (MathisGD) [inline test/forge/HalmosTest.sol:71]: creating a new market doesn't break rules, just break the prover right (timeout)?

File: test/forge/HalmosTest.sol [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        morpho.createMarket(marketParams);
+        vm.stopPrank();
+
+        vm.roll(block.number + 1);
+        vm.warp(block.timestamp + 1 * BLOCK_TIME);
+    }
+
+    // Call Morpho, assuming interacting with only the defined market for performance reasons.
+    function _callMorpho(bytes4 selector, address caller) internal {
+        vm.assume(selector != morpho.extSloads.selector);
```
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol:70]: but actually we don't have problems with functions that always revert no? for example, I assume that authorization functions can't be called without reverting

File: test/forge/HalmosTest.sol
Comment (QGarchery) [reply test/forge/HalmosTest.sol:1]: I have the same warning, I'll ask why

[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:1]: We could even put it in a separate folder, like we do for `certora`. What do you think is the best ?

File: test/forge/HalmosTest.sol [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        irm = new IrmMock();
+
+        vm.startPrank(OWNER);
+        morpho.enableIrm(address(0));
+        morpho.enableIrm(address(irm));
+        morpho.enableLltv(0);
+        morpho.setFeeRecipient(FEE_RECIPIENT);
+        vm.stopPrank();
+
+        uint256 lltv = svm.createUint256("lltv");
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:54]: Yes, they are incompatible now. The name of the test function is different: either start with `test` for forge tests, or start with `check_` for halmos

File: test/forge/HalmosTest.sol [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        marketParams = MarketParams(address(loanToken), address(collateralToken), address(oracle), address(irm), lltv);
+        id = marketParams.id();
+
+        vm.assume(block.timestamp != 0);
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:58]: Yes it has to be symbolic otherwise there would be no valid path of execution. This is because if it was not symbolic it would be at the default `0` value, and this would make it trivial with this line.
If it's zero then you would have a created market with `lastUpdate = 0`, so the rule `check_lastUpdateNonZero` would fail

File: test/forge/HalmosTest.sol [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        loanToken = new ERC20Mock();
+        collateralToken = new ERC20Mock();
+        oracle = new OracleMock();
+        oracle.setPrice(ORACLE_PRICE_SCALE);
+        irm = new IrmMock();
+
+        vm.startPrank(OWNER);
+        morpho.enableIrm(address(0));
+        morpho.enableIrm(address(irm));
+        morpho.enableLltv(0);
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:50]: We can remove it, I copied the setup from the tests and it made sense because this is how morpho is deployed

File: test/forge/HalmosTest.sol [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        morpho.enableLltv(lltv);
+        morpho.createMarket(marketParams);
+        vm.stopPrank();
+
+        vm.roll(block.number + 1);
+        vm.warp(block.timestamp + 1 * BLOCK_TIME);
+    }
+
+    // Call Morpho, assuming interacting with only the defined market for performance reasons.
+    function _callMorpho(bytes4 selector, address caller) internal {
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:69]: Exactly

File: test/forge/HalmosTest.sol [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                token = address(otherToken);
+            }
+            args = abi.encode(marketParams, token, assets, emptyData);
+        } else if (selector == morpho.accrueInterest.selector) {
+            args = abi.encode(marketParams);
+        } else if (selector == morpho.setFee.selector) {
+            uint256 newFee = svm.createUint256("newFee");
+            args = abi.encode(marketParams, newFee);
+        } else {
+            args = svm.createBytes(1024, "data");
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:115]: This snippet of code is here for 2 reasons:
- to disambiguate dynamic types as arguments of symbolic functions, because Halmos does not support it
- to make it easier on the prover, notably here we pass exactly the marketParam for the market that was created in the setup

File: test/forge/HalmosTest.sol [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            uint256 rand = svm.createUint256("rand");
+            address token;
+            if (rand == 0) {
+                token = address(loanToken);
+            } else if (rand == 1) {
+                token = address(collateralToken);
+            } else {
+                ERC20Mock otherToken = new ERC20Mock();
+                token = address(otherToken);
+            }
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:107]: Here we assume that all the other addresses that it wants to use for the token are corresponding to `ERC20Mock` contract. So it restricts the other addresses, other than the loan token and the collateral token. It's like a `DISPATCH` in certora: it's not sound, but at least we can use it. We could improve it, by making so the token can renter Morpho, but maybe improving other areas is more important if we allow it to take more time to compute (like permitting multiple markets)

File: test/forge/HalmosTest.sol [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+    // Check that the fee is always smaller than the max fee.
+    function check_feeInRange(bytes4 selector, address caller) public {
+        _callMorpho(selector, caller);
+
+        assert(morpho.fee(id) <= MAX_FEE);
+    }
+
+    // Check that there is always less borrow than supply on the market.
+    function check_borrowLessThanSupply(bytes4 selector, address caller) public {
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:131]: You can't call a symbolic address, it's not supported by Halmos yet. A default behavior to be able to call back any known contract was considered, but it showed to be very resource intensive

File: test/forge/HalmosTest.sol [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            args = abi.encode(marketParams, assets, onBehalf, emptyData);
+        } else if (selector == morpho.withdrawCollateral.selector) {
+            args = abi.encode(marketParams, assets, onBehalf, receiver);
+        } else if (selector == morpho.liquidate.selector) {
+            address borrower = svm.createAddress("borrower");
+            args = abi.encode(marketParams, borrower, assets, shares, emptyData);
+        } else if (selector == morpho.flashLoan.selector) {
+            uint256 rand = svm.createUint256("rand");
+            address token;
+            if (rand == 0) {
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:100]: Yes, I agree that it feels weird at the beginning, but for symbolic execution it gives exactly the same thing :)

File: test/forge/HalmosTest.sol [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    // Check that the market cannot be "destroyed".
+    function check_lastUpdatedNonZero(bytes4 selector, address caller) public {
+        _callMorpho(selector, caller);
+
+        assert(morpho.lastUpdate(id) != 0);
+    }
+
+    // Check that enabled LLTVs are necessarily less than 1.
+    function check_lltvSmallerThanWad(bytes4 selector, address caller, uint256 lltv) public {
+        vm.assume(!morpho.isLltvEnabled(lltv) || lltv < 1e18);
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol]: `lltv` is symbolic in this example, so it could be another one than the one that was defined in the setup. Also the `isLltvEnabled` mapping is symbolic. 
So it does not prevent the case where `!morpho.isLltvEnalbled(lltv)` and `lltv >= 1e18`. Assuming that we call any function that does not touch this `lltv`, for example a view function, then we would still have `!morpho.isLltvEnalbled(lltv)` and `lltv >= 1e18` after the function call and the assert would fail

File: test/forge/HalmosTest.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol]: Removed

File: test/forge/HalmosTest.sol (hunk 1 above)
Comment (QGarchery) [reply test/forge/HalmosTest.sol:63]: It does break the rules, see [this comment](https://github.com/morpho-org/morpho-blue/pull/675#discussion_r1574685906)

Comment (QGarchery) [reply test/forge/HalmosTest.sol:63]: `createMarket` will fail if we pass symbolic values to it. This is because `createMarket` calls the IRM, and calling a symbolic address is not permitted. Because of this, I found it cleaner to avoid calling `createMarket` altogether

File: test/forge/HalmosTest.sol [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        vm.prank(caller);
+        (bool success,) = address(morpho).call(abi.encodePacked(selector, args));
+        vm.assume(success);
+    }
+
+    // Check that the fee is always smaller than the max fee.
+    function check_feeInRange(bytes4 selector, address caller) public {
+        _callMorpho(selector, caller);
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:112]: I have not tried to do a loop, but here the whole state is symbolic, see the `--symbolic-storage` flag at the top of the test contract definition. It means that it proves the properties by starting in an arbitrary state, and in theory it is enough to show that this is an invariant. It's not entirely true though, because there are some variables that are not symbolic, such as the market that is created in the setup. But it should still give us a very good verification work for those properties

File: test/forge/HalmosTest.sol (hunk 1 above)
Comment (MathisGD) [reply test/forge/HalmosTest.sol:63]: but actually a lot of calls that we do fail and the prover is ok with that no?

File: test/forge/HalmosTest.sol
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol:1]: hmm idk, we could also group them under a formal-verif folder (clean to have src, test, formal-verif)

File: test/forge/HalmosTest.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol]: this idea was cool but kind of unrealistic anyway

File: test/forge/HalmosTest.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol:112]: ok, I completely missed this. Let me update my checks then

[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol:112]: why did they write this article then ?
>Have you tried just looping like they do [here](https://a16zcrypto.com/posts/article/implementing-stateful-invariant-testing-with-halmos/) ?

File: test/forge/HalmosTest.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:62]: The issue here is that you can't access symbolic storage slots, so checking this function is essentially blocked until we can do it
Screenshot_2024-04-22_15-24-24

File: test/forge/HalmosTest.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol:118]: I don't think that changing the id here calls any symbolic address (just reads totalSupply and totalBOrrow for this market)

File: test/forge/HalmosTest.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (MathisGD) [reply test/forge/HalmosTest.sol]: I didn't know that the storage was completely symbolic, that's why I asked this

File: test/forge/HalmosTest.sol (hunk 1 above)
Comment (MathisGD) [reply test/forge/HalmosTest.sol:63]: i understand now, calls to arbitrary addresses are not supported

Comment (QGarchery) [reply test/forge/HalmosTest.sol:63]: Yes but there are 2 kinds of failures: either a revert or a unsupported situation by Halmos. In the latter, it just reports as the test failing. For reverts, Halmos assumes that the path is not taken. This is similar to what Certora is doing (except if you add the `@withrevert` suffix to the function call). This means that you can verify functions that revert for some paths, and verify the property for all of the paths that do not revert.
Here calling a symbolic address is unsupported by Halmos, so the test would always show a fail

File: test/forge/HalmosTest.sol [hunk 13]
```diff
@@ -182,7 +173,7 @@ contract HalmosTest is SymTest, Test {
         _callMorpho(selector, caller);
 
         uint256 nonceAfter = morpho.nonce(user);
-        assert(nonceAfter == nonceBefore || nonceAfter == nonceBefore + 1);
+        assert(nonceAfter == nonceBefore);
```
Comment (MathisGD) [inline test/forge/HalmosTest.sol]: why did you change this? it cannot pass like that right?

File: test/forge/HalmosTest.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:112]: > why did they write this article then ?

I'm assuming it's because they where talking about the usual case where you don't give the flag `--symbolic-storage`. Also there is the caveat that I mentioned about the storage not being completely symbolic

File: test/forge/HalmosTest.sol (hunk 13 above)
Comment (MathisGD) [reply test/forge/HalmosTest.sol]: fixed in da3ebee

File: test/forge/HalmosTest.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:118]: Ah yes, you are right

File: test/forge/HalmosTest.sol (hunk 13 above)
Comment (QGarchery) [reply test/forge/HalmosTest.sol]: Oops my bad, I wanted to look at the error message and I committed this by accident

File: test/forge/HalmosTest.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol]: Done

File: test/forge/HalmosTest.sol
[the remark this answers is not in the document]
Comment (QGarchery) [reply test/forge/HalmosTest.sol:1]: I put them in `test/halmos` since they look a lot like solidity tests. Put we can revert and move them elsewhere if needed

Comment (QGarchery) [reply test/forge/HalmosTest.sol:1]: They are aware of the warning and will remove it in a future version. Nothing to do on our side

File: test/forge/HalmosTest.sol [hunk 14]
```diff
@@ -49,8 +44,7 @@ contract HalmosTest is SymTest, Test {
         marketParams = MarketParams(address(loanToken), address(collateralToken), address(oracle), address(irm), lltv);
         id = marketParams.id();
 
-        vm.assume(block.timestamp != 0);
```
Comment (MathisGD) [inline test/forge/HalmosTest.sol]: in the end, this is not needed?

**Review by MathisGD — APPROVED** [review]

File: test/forge/HalmosTest.sol (hunk 14 above)
Comment (QGarchery) [reply test/forge/HalmosTest.sol]: I changed the function `check_lastUpdateNonZero` to verify for an arbitrary `id`. To do that I needed to assume the precondition of the invariant which makes the line you are highlighting useless in the end. We still rely on the fact that markets start with non zero timestamp, but it's not explicitely written in the verification

**Review by MerlinEgalite — APPROVED** [review]
